### PR TITLE
fix #16: replace bad write-access to netdev addr

### LINKED ISSUE #16: Illegally overwritten netdev address
My colleague @mirage83 also noted another error where the netdev Address is directly written instead of using the proper abstractions to do so. Here is a matching trace for that:
Current addr:  ad d6 bf ec 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
nterface down
```
[   95.784717] irda0 (unregistered): Expected addr: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
[   95.786492] ------------[ cut here ]------------
[   95.787095] netdevice: irda0 (unregistered): Incorrect netdev->dev_addr
[   95.788900] WARNING: CPU: 0 PID: 487 at net/core/dev_addr_lists.c:519 dev_addr_check+0x94/0x140
[   95.790013] Modules linked in: irtty_sir(O) sir_dev(O) irda(O) hbm_fpga sec_mipi_dsim_imx sec_dsim panel_hbm_mipi spi_nor panel_hbm imx_sdma virt_dma
[   95.791723] CPU: 0 PID: 487 Comm: hbm_app Tainted: G           O       6.1.55-[10.00.00]-hbm+ #69
[   95.792841] Hardware name: Roche HBM5 board with production BSP (DT)
[   95.793642] pstate: 60000005 (nZCv daif -PAN -UAO -TCO -DIT -SSBS BTYPE=--)
[   95.794521] pc : dev_addr_check+0x94/0x140
[   95.795040] lr : dev_addr_check+0x94/0x140
[   95.795560] sp : ffff80000873ba40
[   95.795979] x29: ffff80000873ba40 x28: ffff1879044e8000 x27: 0000000000000000
[   95.796886] x26: ffff18790468b000 x25: 0000ffff91c1ce7c x24: ffff18790468b000
[   95.797790] x23: 0000000000000000 x22: ffff1879046a4000 x21: 0000000000000122
[   95.798695] x20: ffff1879046a4000 x19: ffff1879046a4000 x18: 0000000000000000
[   95.799598] x17: 0000000000000001 x16: 00001b8f812b2fc4 x15: 006e521efc3a345e
[   95.800504] x14: ffffceedcb8fb138 x13: 00000000000000cb x12: 00000000fa83b2da
[   95.801408] x11: 00000000005bbe78 x10: 00000000000007e0 x9 : ffff80000873b900
[   95.802314] x8 : ffff1879044e8840 x7 : 0000000000000001 x6 : 000000164c417000
[   95.803218] x5 : 0000000000000000 x4 : 0000000000000000 x3 : 0000000000000001
[   95.804120] x2 : 0000000000000000 x1 : 0000000000000000 x0 : ffff1879044e8000
[   95.805026] Call trace:
[   95.805337]  dev_addr_check+0x94/0x140
[   95.805815]  dev_addr_flush+0x24/0x8c
[   95.806280]  free_netdev+0x78/0x1a0
[   95.806725]  sirdev_put_instance+0x84/0xc0 [sir_dev]
[   95.807359]  irtty_close+0x38/0xfc [irtty_sir]
[   95.807928]  tty_ldisc_close+0x38/0x80
[   95.808408]  tty_set_ldisc+0xdc/0x1fc
[   95.808875]  tty_ioctl+0x49c/0xa00
[   95.809310]  __arm64_sys_ioctl+0xa8/0xf0
[   95.809809]  invoke_syscall+0x70/0xf4
[   95.810279]  el0_svc_common.constprop.0+0x44/0xfc
[   95.810875]  do_el0_svc+0x30/0xd0
[   95.811297]  el0_svc+0x2c/0x84
[   95.811688]  el0t_64_sync_handler+0xbc/0x140
[   95.812229]  el0t_64_sync+0x18c/0x190
[   95.812694] ---[ end trace 0000000000000000 ]---
```

I believe I have a matching MR for that. Going to post shortly.

### PR DESCRIPTION
**Pull request by dpiegdon**
(not tested yet)

### CODE REVIEW

File: src/net/irlap.c [hunk 1]
```diff
@@ -147,7 +154,13 @@ struct irlap_cb *irlap_open(struct net_device *dev, struct qos_info *qos,
 	} while ((self->saddr == 0x0) || (self->saddr == BROADCAST) ||
 		 (hashbin_lock_find(irlap, self->saddr, NULL)) );
 	/* Copy to the driver */
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 15, 0)
 	memcpy((void *)dev->dev_addr, &self->saddr, 4);
+#else
+	memset(eth_addr, 0, sizeof(eth_addr));
+	memcpy(eth_addr, &self->saddr, MIN(sizeof(eth_addr), sizeof(self->saddr)));
+	eth_hw_addr_set(dev, eth_addr);
```
Comment (cschramm) [inline src/net/irlap.c]: `eth_hw_addr_set` is the wrong abstraction here.

I think this should rather be just `__dev_addr_set(dev, &self->saddr, LAP_ALEN);`

File: src/net/irlap.c [hunk 2]
```diff
@@ -147,7 +154,13 @@ struct irlap_cb *irlap_open(struct net_device *dev, struct qos_info *qos,
 	} while ((self->saddr == 0x0) || (self->saddr == BROADCAST) ||
 		 (hashbin_lock_find(irlap, self->saddr, NULL)) );
 	/* Copy to the driver */
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 15, 0)
```
Comment (cschramm) [inline src/net/irlap.c]: 5.15 is the current minimum target, so this is unnecessary.

**Review by cschramm — CHANGES_REQUESTED** [review]

File: src/net/irlap.c (hunk 1 above)
Comment (dpiegdon) [reply src/net/irlap.c]: agreed. the latest change uses dev_addr_set(), which uses dev->addr_len. that should have been set to LAP_ALEN, so just BUG() if that's not the case.

**Comment by mirage83** [comment]
I tested this patch. It resolves the observed issue.

**Review by mirage83 — APPROVED** [review]

**Review by cschramm — APPROVED** [review]
Awesome, thanks! 👏

**Pull request merged**